tensorflow_io/arrow/kernels/arrow_dataset_ops.cc

```diff
@@ -0,0 +1,615 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "arrow/api.h"
+#include "arrow/adapters/tensorflow/convert.h"
+#include "arrow/ipc/api.h"
+#include "arrow/util/io-util.h"
+#include "tensorflow_io/arrow/kernels/arrow_stream_client.h"
+#include "tensorflow/core/framework/dataset.h"
+
+#define CHECK_ARROW(arrow_status)             \
+  do {                                        \
+    arrow::Status _s = (arrow_status);        \
+    if (!_s.ok()) {                           \
+      return errors::Internal(_s.ToString()); \
+    }                                         \
+  } while (false)
+
+namespace tensorflow {
+
+// Convert an element of an Arrow Array to a Tensor
+class ArrowConvertTensor : public arrow::ArrayVisitor {
+ public:
+  ArrowConvertTensor(int64_t row_idx, IteratorContext* ctx)
+      : curr_row_idx_(row_idx), curr_ctx_(ctx), curr_array_values_(-1) {}
+
+  // Convert to a Tensor and append to the output vector
+  Status AppendTensor(std::shared_ptr<arrow::Array> array, DataType output_type,
+                      std::vector<Tensor>* out_tensors) {
+    curr_type_ = output_type;
+    out_tensors_ = out_tensors;
+    if (array->null_count() != 0) {
+      return errors::Internal("Arrow arrays with null values not supported");
+    }
+    CHECK_ARROW(array->Accept(this));
+    return Status::OK();
+  }
+
+ protected:
+  template <typename ArrayType>
+  arrow::Status VisitFixedWidth(const ArrayType& array) {
+
+    // TODO only need to check once per stream of batches
+    DataType converted_type;
+    ARROW_RETURN_NOT_OK(arrow::adapters::tensorflow::GetTensorFlowType(
+        array.type(), &converted_type));
+    if (curr_type_ != converted_type) {
+      return arrow::Status::TypeError(
+          "Type mismatch from Arrow batch for expected dtype=" +
+          std::to_string(curr_type_) + ", but got dtype=" +
+          std::to_string(converted_type));
+    }
+
+    const auto& fw_type =
+        static_cast<const arrow::FixedWidthType&>(*array.type());
+    const int64_t type_width = fw_type.bit_width() / 8;
+
+    // Create a Tensor of required size
+    // curr_array_values_ < 0 indicates scalar
+    Tensor tensor(curr_ctx_->allocator({}), curr_type_,
+                  curr_array_values_ < 0 ? TensorShape({})
+                                         : TensorShape({curr_array_values_}));
+
+    auto values = array.data()->buffers[1];
```

Comment:
sorry i'm not so familiar with arrow, but why is `buffers[1]` here?

Reply:
No prob, it's not clear in this context. For primitive types, the first buffer is a validity bitmap to indicate NULL values, and the second buffer is the data values. 

There is a check to make sure NULL count is zero here https://github.com/tensorflow/io/pull/36/files#diff-42f74bbc07801dbac60e26f2d9fd6f70R44, so we don't care about that first buffer (for now at least)

I will make this a `static const int VALUE_BUFFER = 1` and add a note to make it more clear